xalia/Win32/NonclientScrollProvider.cs

```diff
@@ -409,5 +409,87 @@ internal void MsaaValueChange()
             else
                 _siKnown = false;
         }
+
+        public async Task<double> GetMinimumIncrementAsync(UiDomElement element)
+        {
+            try
+            {
+                return await Connection.CommandThread.OnBackgroundThread(() =>
+                {
+                    var sbi = new SCROLLBARINFO();
+                    sbi.cbSize = Marshal.SizeOf<SCROLLBARINFO>();
+                    if (!GetScrollBarInfo(Hwnd, Vertical ? OBJID_VSCROLL : OBJID_HSCROLL, ref sbi))
+                        throw new Win32Exception();
+
+                    var si = new SCROLLINFO();
+                    si.cbSize = Marshal.SizeOf<SCROLLINFO>();
+                    si.fMask = SIF_PAGE;
+                    if (!GetScrollInfo(Hwnd, Vertical ? SB_VERT : SB_HORZ, ref si))
+                        throw new Win32Exception();
+
+                    return CalculateMinimumIncrement(sbi, si);
+                }, Tid);
+            }
+            catch (Win32Exception e)
+            {
+                if (!HwndProvider.IsExpectedException(e))
+                    throw;
+                return 1.0;
+            }
+        }
+
+        public async Task SetValueAsync(int value)
+        {
+            int msg = Vertical ? WM_VSCROLL : WM_HSCROLL;
+            await SendMessageAsync(Hwnd, msg, MAKEWPARAM(SB_THUMBTRACK, unchecked((ushort)value)), IntPtr.Zero);
+            await SendMessageAsync(Hwnd, msg, MAKEWPARAM(SB_THUMBPOSITION, unchecked((ushort)value)), IntPtr.Zero); ;
+            await SendMessageAsync(Hwnd, msg, MAKEWPARAM(SB_ENDSCROLL, 0), IntPtr.Zero);
+        }
+
+        double _offsetRemainder;
+
+        public async Task<bool> OffsetValueAsync(UiDomElement element, double offset)
+        {
+            try
+            {
+                SCROLLINFO si = await Connection.CommandThread.OnBackgroundThread(() =>
+                {
+                    var bg_si = new SCROLLINFO();
+                    bg_si.cbSize = Marshal.SizeOf<SCROLLINFO>();
+                    bg_si.fMask = SIF_POS | SIF_PAGE | SIF_RANGE;
+                    if (!GetScrollInfo(Hwnd, Vertical ? SB_VERT : SB_HORZ, ref bg_si))
+                        throw new Win32Exception();
+
+                    return bg_si;
+                }, Tid);
+
+                if (si.max_value <= si.nMin)
+                    return false;
+
+                double new_pos = si.nPos + offset + _offsetRemainder;
+
+                if (new_pos < si.nMin)
+                    new_pos = si.nMin;
+                else if (new_pos > si.max_value)
+                    new_pos = si.max_value;
+
+                int new_pos_int = (int)Math.Round(new_pos);
+
+                if (new_pos_int != si.nPos)
+                {
+                    await SetValueAsync(new_pos_int);
+                }
+
+                _offsetRemainder = new_pos - new_pos_int;
+

```

Comment:
What happens if the scrollbar is moved in the background (e.g not by Xalia), would this value be incorrect the next time this method is called?

Reply:
The threshold before the scrollbar moves a single unit would be smaller (or larger if moving in the other direction). Arguably, perhaps the control should slowly "recenter" over time, but in practice I haven't noticed an issue. Usually, even if a single unit is a full line, it's small enough that in practice all this really does is "smooth out" the motion.